mtt-management/src/mtt_types.rs

```diff
@@ -137,3 +133,189 @@ pub enum Error {
+    let range_start = PageSize::Size1G.round_up(phys_addr.bits());
+    let first_entry_type = get_l2_entry(range_start);
+    // Enforce the invariant that all 16 L2 entries for the 1GB range must have the same type
+    for i in 0u64..=15 {
+        let addr = range_start + (i * 64 * 1024 * 1024);
+        let mtt_entry = get_l2_entry(addr);
+        match mtt_entry {
+            MttNonConfidential1G if mtt_entry == first_entry_type => {
+                set_l2_entry(addr, MttL2EntryType::non_confidential_64_mb_range())
+            }
+            MttConfidential1G if mtt_entry == first_entry_type => {
+                set_l2_entry(addr, MttL2EntryType::confidential_64_mb_range())
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+// Changes the L2 type of 64MB range containing the address to 4Kx16KB.
+// All pages in the range will retain the original MTT type (non-confidential or confidential).
+// The `mtt_l1_page` is supplied by the caller and will contain 2-bit encoding for every page
+// in the fractured range. The caller must ensure that the page isn't freed until there are
+// no outstanding references to it in the L2 table.
+fn fracture_64mb_range(phys_addr: SupervisorPageAddr, mtt_l1_page: SupervisorPageAddr) {
+    let addr = mtt_l1_page.bits();
+    // Safety: mtt_l1_page is a supervisor page address that's owned by us, and is at least 4K in size
+    unsafe {
+        core::ptr::write_bytes(addr as *mut u8, 0, PageSize::Size4k as usize);
+    }
+    match get_l2_entry(phys_addr.bits()) {
+        Mtt2MBPages(value) => {
+            for i in 0u64..=31 {
+                // If the 64MB range is confidential, mark it in the 4K L2 page
+                let byte_value = if value & (1 << 1) == 0 { 0u8 } else { 0x55u8 };
+
+                // Safety: mtt_l1_page is a supervisor page address that's owned by us, and is at least 4K in size
+                // The maximum index is 31*128 from the base and we write 128-bytes at a time
+                unsafe {
+                    core::ptr::write_bytes((addr + (i * 128)) as *mut u8, byte_value, 128);
+                }
+            }
+
+            set_l2_entry(phys_addr.bits(), MttL2EntryType::mtt_l1_dir(mtt_l1_page));
+        }
+        _ => unreachable!(),
+    }
+}
+
+/// Sets the type of the 4K page corresponding to `phys_addr` to the type specified in `memory_type` (non-confidential or confidential).
+/// The caller is responsible for allocating a 4K page if there's no previously-allocated L1 page corresponding to the address. The function
+/// function will fail if `mtt_l1_page` is None, and the L2 entry for the page doesn't have an entry for L1. The caller must ensure that
+/// deallocation of the L1 page happens only when there are no outstanding references to it in the L1 table.
+pub fn set_4k_page_type(
+    phys_addr: SupervisorPageAddr,
+    memory_type: MttMemoryType,
+    mtt_l1_page: Option<SupervisorPageAddr>,
+) -> Result<()> {
```

Comment:
It seems like your suggestion is to always pass in a 4K L1 page, and then return it in the Result<> if it's not consumed? I will have to think more about it, but it might be easier for the caller to simply track whether a 4K page is allocated for the 64MB region and pass it in (and use the tracking logic to free it). 

BTW, there's no existing code to call these routines (yet), so suggestions are welcome.,

Reply:
I was thinking to leave the Option<> parameter and error behavior in place for cases where it's obvious that no page is needed. But it was really just a suggestion, we should do whatever is most ergonomic for the calling code.

Reply:
Based on offline discussions, we decided that the best approach was to simply reserve a page pool for the entire physical address space upfront.

1. At 16 4K-pages per 1GB (one page per 64MB), this translates into a fairly modest 384MB assuming 6TB of platform memory
2. It obviates the need to create new APIs between the host-VMM and the TSM to create and manage the pool for 4K pages
3. The fixed mapping reduces the possibility of errors because there will be fixed mapping between the physical address and the associated 4K page in the pool.

Reply:
I do like the simplicity of the proposed approach, and it'll probably be good enough for the first iteration.

On the other hand, 384MB/6TB is 6% proportional memory overhead, and in my previous live I've seen cases where folks have been investing significant time into avoiding much less significant memory tax situations...